spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/ErrorListenerAnnotationBeanPostProcessor.java

```diff
@@ -0,0 +1,97 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.BeanPostProcessor;
+import org.springframework.cloud.stream.annotation.ErrorListener;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.annotation.AnnotatedElementUtils;
+import org.springframework.integration.channel.PublishSubscribeChannel;
+import org.springframework.integration.context.IntegrationContextUtils;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.core.DestinationResolver;
+import org.springframework.messaging.handler.annotation.support.MessageHandlerMethodFactory;
+import org.springframework.messaging.handler.invocation.InvocableHandlerMethod;
+import org.springframework.util.ReflectionUtils;
+
+/**
+ * A {@link BeanPostProcessor} that handles {@link ErrorListener} annotations found on bean methods.
+ *
+ * @author Ilayaperumal Gopinathan
+ */
+public class ErrorListenerAnnotationBeanPostProcessor implements BeanPostProcessor, ApplicationContextAware {
+
+	@Autowired
+	@Lazy
+	private DestinationResolver<MessageChannel> binderAwareChannelResolver;
+
+	@Autowired
+	@Lazy
+	private MessageHandlerMethodFactory messageHandlerMethodFactory;
+
+	private ConfigurableApplicationContext applicationContext;
+
+	@Override
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	public final void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = (ConfigurableApplicationContext) applicationContext;
+	}
+
+	@Override
+	public final Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+		return bean;
+	}
+
+	@Override
+	public final Object postProcessAfterInitialization(final Object bean, final String beanName) throws BeansException {
+		Class<?> targetClass = AopUtils.isAopProxy(bean) ? AopUtils.getTargetClass(bean) : bean.getClass();
+		ReflectionUtils.doWithMethods(targetClass, new ReflectionUtils.MethodCallback() {
+			@Override
+			public void doWith(final Method method) throws IllegalArgumentException, IllegalAccessException {
+				ErrorListener errorListener = AnnotatedElementUtils.findMergedAnnotation(method,
+						ErrorListener.class);
+				if (errorListener != null && !method.isBridge()) {
+					registerHandlerMethodOnListenedChannel(method, errorListener, bean);
+				}
+			}
+		});
+		return bean;
+	}
+
+	protected final void registerHandlerMethodOnListenedChannel(Method method, ErrorListener errorListener,
+			Object bean) {
+		Method targetMethod = StreamListenerMethodUtils.checkProxy(method, bean);
+		final InvocableHandlerMethod invocableHandlerMethod = this.messageHandlerMethodFactory
+				.createInvocableHandlerMethod(bean, targetMethod);
```

Comment:
Well, since you rely on the `messageHandlerMethodFactory`, we really can deal just with the exception `payload` and even `@Header` and some other tricks. See `DefaultMessageHandlerMethodFactory.initArgumentResolvers()`.